Use Mach time  when clock_gettime() not suppported

sveljko opened the pull request:
Older versions of MacOS did not support POSIX `clock_gettime()`, but newer do.
So, we have a runtime check to see if we're running on a MacOS that does support
it, and use the Mach kernel time APIs for older MacOSes. This does incur a runtime
check cost, but is a clear path forward for when the support for older MacOSes is
dropped.

There are some "for review" comments which are intended to be removed after
review is done, you'll discover then quickly. :)

After pondering for a while whether to somehow integrate this into the build/CMake,
I opted to leave it as a "code only" change with some macro magic... This does
make it a little ugly as we have to repeat the code that uses `clock_gettime()` twice,
but it's actually trivial code, which is done in 4 lines here, for whatever reason... Trying
to avoid that would make for even uglier macro trickery, and that's never good.

sveljko committed:
Use Mach time  when clock_gettime() not suppported

Older versions of MacOS did not support POSIX `clock_gettime()`, but newer do.
So, we have a runtime check to see if we're running on a MacOS that does support
it, and use the Mach kernel time APIs for older MacOSes. This does incur a runtime
check cost, but is a clear path forward for when the support for older MacOSes is
dropped.

SergeyKrasnozhon commented:
Source - https://tasks.hubstaff.com/app/organizations/55/projects/2221/tasks/3392469

SergeyKrasnozhon commented on src/platform/posix/posix_clock.c:
```diff
@@ -19,13 +19,66 @@
 	nni_time        msec;
 
+#if __APPLE__
+
```
[minor; thoughts only]

how crazy would it be if we replace duplicated code with `goto nni_clock_original ` which is a line 82 here?
so it is a `if (__builtin_available(macOS 10.12, *)) { goto nni_clock_original }` in which case we don't need duplications and that final #else. 

[not important] And maybe encapsulate the pre-osx12 code into `nni_clock_osx12_compat(...)`

So the final result is like
```
#ifdef Apple checks
       if OSX_12 then goto nni_clock_original
       return nni_clock_osx12_compat(...)
nni_clock_original:
#endif
	if (clock_gettime(NNG_USE_CLOCKID, &ts) != 0) {
 		// This should never ever occur.
 		nni_panic("clock_gettime failed: %s", strerror(errno));
	}
	msec = ts.tv_sec;
 	msec *= 1000;
 	msec += (ts.tv_nsec / 1000000);
 	return (msec);
```

eakoli replied:
I think it can be even simpler than that

```
nni_clock(void)
 {
 	struct timespec ts;
 	nni_time        msec;

+#if defined( __APPLE__ )
+      // if the symbol is not available it will be null
+.     if( clock_gettime == 0 ) {
+.         /// alternate implementation of clock_gettime
+.    } else
+#endif
        if (clock_gettime(NNG_USE_CLOCKID, &ts) != 0) {
 		// This should never ever occur.
 		nni_panic("clock_gettime failed: %s", strerror(errno));
	}
	msec = ts.tv_sec;
 	msec *= 1000;
 	msec += (ts.tv_nsec / 1000000);
 	return (msec);
}
```

sveljko replied:
I really dislike `goto`s.  I guess it comes from having written some assembly in my day and seeing how easy it is to abuse... I mean, there are some cool things, like ad-hoc exception handling, but, boy is it fragile! :)

We could make a helper macro or inline function for `clock_gettime() stuff` and then use it twice, to avoid code duplication.

I'm also "allergic" to playing tricks with standard libraries symbols. There's too much of a chance of there being some macro or "intrinsic" compiler magic in there.

eakoli replied:
Well to reduce any duplicated code the basic flow is correct.

```
    ...
#if defined(__APPLE__)
    if( <detect alt implementation needed > ) {
          // alternate implementation
    } else
#endif
  ...
```

There are two main options for how to handle the if case
a) always use the alternate if __MAC_OS_X_VERSION_MIN_REQUIRED < __MAC_10_12
b) conditionally use the alternate if the runtime check < 10.12

sveljko replied:
I did the `static inline` common code optimisation :)

SergeyKrasnozhon replied:
[2 cents; still approved]

I personally prefer the
```
#if defined(__APPLE__)
    if( <detect alt implementation needed > ) {
          // alternate implementation
    } else
#endif
```
- this looks and feels like a ox12 patch, not the whole function refactoring. so it is encapsulated and can be easily removed. sure, not always this would be possible, so in general I am not criticizing your approach
- after this refactoring, we really "added a logic" that people now need to read and understand. while with the patch it is like "oh, I am not on apple, I can ignore"
- with the patch, the intention seems to be written clearer: "the main flow remains the same. but here is a special condition that will do something alternative"

sveljko replied:
The problem is that it's a little more involved to do that `<detect alt implementation>`, and the code looks a little stranger (and with a hanging `else` in the `#if` we would need to add a `{` for the rest and then another `#if` to close it...

But, after some more playing with it, I came up with what's in the latest version. It's a little involved, but only on `APPLE`, on other platforms, there's no change at all.

Quite frankly, if this was for our use only, I'd prefer the previous version. Refactoring is really minimal and code looks "more regular". But, if we're to push the patch upstream, this version is much more likely to be accepted.

SergeyKrasnozhon replied:
this last version looks cool to me!

eakoli replied:
the handling else isn't really an issue, as the intention would be your just replacing the clock_gettime call, not reimplementing the nni_clock() method. That is the alt code ONLY initializes the `struct timespec ts`, thus the final logic is all the same.

thus on non apple the code would be 

```
nni_time
 nni_clock(void)
 {
 	struct timespec ts;
 	nni_time        msec;

<-- if def removes all code -->

	if (clock_gettime(NNG_USE_CLOCKID, &ts) != 0) {
 		// This should never ever occur.
 		nni_panic("clock_gettime failed: %s", strerror(errno));
	}
	msec = ts.tv_sec;
	msec *= 1000;
	msec += (ts.tv_nsec / 1000000);
	return (msec);
}
```

and on apple it would be 

```
nni_time
 nni_clock(void)
 {
 	struct timespec ts;
 	nni_time        msec;

        bool use_alt = ....
        if( use_alt ) {
           // initialize ts
        } 
        else 
	if (clock_gettime(NNG_USE_CLOCKID, &ts) != 0) {
 		// This should never ever occur.
 		nni_panic("clock_gettime failed: %s", strerror(errno));
	}
	msec = ts.tv_sec;
	msec *= 1000;
	msec += (ts.tv_nsec / 1000000);
	return (msec);
}
```

sveljko committed:
static inline the common code

Thus, avoid code duplication. Also removed the slow alternative, it was more of a comment anyway, but it's not worth keeping. Condensed some comments, they are not that important.

eakoli commented on src/platform/posix/posix_clock.c:
```diff
@@ -19,22 +19,57 @@
+nni_time
+nni_clock(void)
+{
+#if __APPLE__
```
```suggestion
#if defined(__APPLE__)
```

you cannot use #if <IDENT> if the identifier could be undefined.

`__APPLE__` is defined only on apple platforms.

sveljko replied:
Not sure what you mean... Try to compile this:
```
#if ZEC
#error ZEC is defined
#endif
```
and it will not report the compiler error (well, unless you define `ZEC` (means rabbit in Serbian), but it's surely not defined by default). Yes it may look a little odd, but it's how people do it... maybe there's some case where someone does
```
#define __APPLE__ 0
```
to indicate a non-Apple platform, which should be handled... I did not go into this.

eakoli replied:
according to apple code snippets the correct check is

```
#ifdef __APPLE__
```
or
```
#if defined(__APPLE__)
```

as its something defined at the compiler level, not at source level. developers should not be re-defining things provided by the compiler. The compiler only states that it will be defined, not what its value will be.

sveljko replied:
I've seen both used in various cross-platform codebases, `#if __APPLE__` seems "broader" as it handles the `#define __APPLE__ 0`, too. But, it's a simple change to go from one to the other, so, I don't care and it seems you do. :)

sveljko committed:
Nicer diff

sveljko committed:
tabs

sveljko committed:
end of line

eakoli commented on src/platform/posix/posix_clock.c:
```diff
@@ -19,13 +19,45 @@
 
+#if __APPLE__
+
+	int supported = 0;
```
for clarity I would use `has_clock_gettime`

eakoli commented on src/platform/posix/posix_clock.c:
```diff
@@ -19,13 +19,45 @@
+	int supported = 0;
+
+#if defined(MAC_OS_X_VERSION_10_12) && __has_builtin(__builtin_available)
+	if (__builtin_available(macOS 10.12, *)) {
```
is it possible (since that evaluates to a boolean type value ) to just use it inline ?

```
bool has_clock_gettime = __builtin_available(macOS 10.12, *);
```

sveljko replied:
Nope, `__builtin_available()` can only be used inside an `if` (I lost a reference for that and it's too late here to look it up again).

eakoli replied:
bummer

eakoli commented on src/platform/posix/posix_clock.c:
```diff
@@ -19,13 +19,45 @@
+
+	int supported = 0;
+
+#if defined(MAC_OS_X_VERSION_10_12) && __has_builtin(__builtin_available)
```
it also might make sense to completely remove the check when its not needed

```
#if (TARGET_OS_OSX && __MAC_OS_X_VERSION_MIN_ALLOWED < 101200)
// include support for clock_gettime fallback as we are targeting pre 10.12
#endif
```

this would remove the need to check for MAC_OS_X_VERSION_10_12, as were not trying to make it compile on older SDKs just be compatible with older OSs.

we also shouldn't need to check that __builtin_available is available, as were only apple which has that feature.

sveljko replied:
Well, a part of the idea was to push this upstream at some point and the thinking is that they would like to see it compile on older SDKs.

sveljko replied:
Also, I've seen the check for `__builtin_available` in many places, I'm guessing it's also to support older SDKs or some such thing, I did not get into it much, it's clear and cheap, so I just let it be.

eakoli replied:
the code as it stands will not compile on < SDK 10.12 due to clock_gettime not even being defined.
I don't think we need to address that, as all were trying to address is the targeting of OS version < 10.12

so im propossing this

```
#if __APPLE__
#if (TARGET_OS_OSX && __MAC_OS_X_VERSION_MIN_ALLOWED < 101200)
    int has_clock_gettime = 0;
    if (__builtin_available(macOS 10.12, *)) {
         has_clock_gettime = 1;
    }
    if( ! has_clock_gettime ) {
        ....
    } else
#endif
#endif
```

sveljko replied:
OK, so it's @SergeyKrasnozhon 's fault, as usual, while making it nicer for him, I also broke the build on older SDKs. Should be fixed now! :)

SergeyKrasnozhon replied:
well... you mentioned something like "Well, a part of the idea was to push this upstream at some point and the thinking is that they would like to see it **compile on older SDKs**." and Chris reasonably mention that even their original code is not going to build on older SDKs :-) 
And also Chris mentioned "I don't think we need to address that", which I agree with :)

sveljko replied:
Their original code can build on older SDKs if you opt to use `gettimeofday()` instead of `clock_gettime()`, but we don't want to do that, cause it's prone to time-traveling problems (adjusting time).

sveljko committed:
supported -> has_clock_gettime

sveljko committed:
#if -> #ifdef

sveljko committed:
another one...

sveljko committed:
Make it build on older SDKs

eakoli commented on src/platform/posix/posix_clock.c:
```diff
@@ -19,13 +19,46 @@
+
+		// mach_continuous_time() is a better option, but it's only                                    
+		// available since MacOS 10.12, for which we already use clock_gettime().                      
+		uint64_t absolute_time = mach_absolute_time();
```
From apple documentation
```
mach_absolute_time
Returns current value of a clock that increments monotonically in tick units (starting at an arbitrary point), this clock **does not increment while the system is asleep**.
```

and for clock_gettime

```
CLOCK_REALTIME     the system's real time (i.e. wall time) clock,
                        expressed as the amount of time since the Epoch.  This
                        is the same as the value returned by [gettimeofday(2)](http://www.manpagez.com/man/2/gettimeofday/).

CLOCK_MONOTONIC    clock that increments monotonically, tracking the time
                        since an arbitrary point, and will continue to incre-
                        ment while the system is asleep.
```

On apple, clock_gettime will be called with CLOCK_REALTIME, so we can just use gettimeofday to implement it.

we already have an issue with macOS sleeping causing issues, and we should match the result based on how the routine was ultimately called (Which is to return a realtime clock)

Additionally based on the NNG code, they provide an alternate path for when clock_gettime is non existent already, just with a big waning.

I suggest we expedite all this and just provide an implementation of clock_gettime directly in our macOS code, and just let other that might need pre-10.12 support deal with It themselves.

sveljko replied:
Not sure what you're aiming for here. The latest version is now complete, it can even build on older SDKs. Unless you have more stylistic remarks. Not sure how doing something completely anew would "expedite things'. 

Also, what is used with `clock_gettime()` can be changed. Currently we do use `CLOCK_REALTIME`, but not everybody does that, AFAICT by looking around for `nng` clones on Github. We probably should change that, but that is a separate matter.

OTOH, in this same file, there is a huge comment against using `gettimeofday()`:

> // If you're here, its because you don't have a modern clock_gettime with
> // monotonic clocks, or the necessary pthread_condattr_settclock().  In
> // this case, you should be advised that *bad* things can happen if your
> // system clock changes time while programs using this library are running.
> // (Basically, timeouts can take longer or shorter, leading to either hangs
> // or apparent spurious errors.

eakoli replied:
What I'm saying is that based on how nng is currently implemented for apple, it calls clock_gettime() using CLOCK_REALTIME, but you implemented the fallback code as though it was called with CLOCK_MONOTONIC, which has two issues : A) its not meeting the contract of how it was called, and B) the monotonic implementation doesn't meet the contract defined for CLOCK_MONOTONIC.

So i'm saying, rather than try and figure out how to formulate a patch that fits in properly with how NNG works, lets just fix OUR problem so that it works the same on all versions of macOS as NNG is implemented today (which was the original scope of the work, which once that was done, we were to look at if it made sense to provide a patch upstream).

in the end, its best not to end up in a situation where we HAVE to maintain a fork.

sveljko replied:
Then I guess something like #2 would be along those lines?

sveljko commented:
Closing this because:
1. We made a "workaround" in our code w/out changing `nng`. Thus we don't have an urgency to finish this.
2. This is not a simple change, even in our team there is some disagreement about what's the best way to go about it. Thus, it will probably not be simple to push it upstream and maintaining a patch can be tedious and time consuming.
3. In some time, we won't support these older MacOS version anyway.

If need be, we can always re-open this. It's a sound and tested solution.
